api/agent/agent.go

```diff
@@ -583,24 +592,54 @@ func (s *hotSlot) exec(ctx context.Context, call *call) error {
 	defer stdoutWrite.Close()
 
 	proto := protocol.New(protocol.Protocol(call.Format), stdinWrite, stdoutRead)
-
 	swapBack := s.container.swap(stdinRead, stdoutWrite, call.stderr, &call.Stats)
 	defer swapBack() // NOTE: it's important this runs before the pipes are closed.
 
 	errApp := make(chan error, 1)
 	go func() {
 		ci := protocol.NewCallInfo(call.Call, call.req)
-		errApp <- proto.Dispatch(ctx, ci, call.w)
+		err := proto.Dispatch(ctx, ci, call.w)
+
+		// drain the container stdout if bad gw (aka container communication failed)
+		if err != nil && models.GetAPIErrorCode(err) == http.StatusBadGateway {
+			go func() {
+				// stdoutWrite.Close() will unblock this with EOF incoming from stdoutRead
+				io.Copy(ioutil.Discard, stdoutRead)
+			}()
+		}
+		errApp <- err
 	}()
 
+	var finalError error
+
 	select {
 	case err := <-s.errC: // error from container
 		return err
 	case err := <-errApp: // from dispatch
-		return err
+		if err == nil || models.GetAPIErrorCode(err) != http.StatusBadGateway {
+			return err
+		}
+		finalError = err
 	case <-ctx.Done(): // call timeout
 		return ctx.Err()
 	}
+
+	// we are here because dispatch IO comm failed. Wait again up to
+	// 500 msecs to drain the I/O pipe to reduce subsequent errors
+	// in hot container. This 500 msec window drain does not necessarily
+	// guarantee that we clean these pipes (as the producer can spit out
+	// more data after) but this does clear naive (most common) cases out
+	// such as container spitting out too much (or faulty) data at once.
+	// If we fail to clear this pipe (stdoutRead), then docker IO demultiplex
+	// stdCopy will notice that we have not read all the data from it and will
+	// shutdown the container.
+	select {
+	case finalError = <-s.errC: // give precedence to error from container
+	case <-time.After(time.Duration(500) * time.Millisecond):
+	case <-ctx.Done(): // call timeout
+	}
```

Comment:
closing the pipes takes care of this and this is redundant I think? the docker stdcopy read will be stuck on an old pipe reader in io.Read? see: conf coffee issue. bbl.

Reply:
So, this has turned out to be tricky. If we close the pipe here, then docker stdcopy doesn't like it. Partial read causes docker to exit the container. That exit 2 event happens after we return here, and sometimes causes next request to fail. Adding 500msec latency to these failures is something I don't like (and it also doesn't 100% cover all cases), but the alternative is to kill the hot container. (which I actually like. :-) )

Reply:
hmm. would this add a condition to our going terminator on containers?

* if container i/o exceeds time bounds
* if container i/o doesn't respect our imposed bounds?

this one seems recoverable, but maybe no